### ISSUE TITLE
Raising ConcurrentModificationException

### ISSUE THREAD

**Issue by zaquestion** [sladkoff/minecraft-prometheus-exporter#3]
Seeing a few of these
```
infinity_1  | [09:51:48 WARN]: 2017-02-17 09:51:48.471:WARN:oejs.HttpChannel:qtp888103540-1642: /metrics
infinity_1  | java.util.ConcurrentModificationException
infinity_1  |   at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901)
infinity_1  |   at java.util.ArrayList$Itr.next(ArrayList.java:851)
infinity_1  |   at org.bukkit.craftbukkit.v1_11_R1.CraftWorld.getLivingEntities(CraftWorld.java:639)
infinity_1  |   at de.sldk.mc.MetricsController.handle(MetricsController.java:35)
infinity_1  |   at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
infinity_1  |   at org.eclipse.jetty.server.Server.handle(Server.java:564)
infinity_1  |   at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:318)
infinity_1  |   at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
infinity_1  |   at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
infinity_1  |   at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:112)
infinity_1  |   at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
infinity_1  |   at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
infinity_1  |   at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
infinity_1  |   at java.lang.Thread.run(Thread.java:745)
```

Only seen 1 of these so far
```
infinity_1  | [10:13:58 WARN]: 2017-02-17 10:13:58.474:WARN:oejs.HttpChannel:qtp888103540-5065: /metrics
infinity_1  | java.lang.NullPointerException
infinity_1  |   at it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap$MapIterator.nextEntry(Long2ObjectOpenHashMap.java:511)
infinity_1  |   at it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap$ValueIterator.next(Long2ObjectOpenHashMap.java:706)
infinity_1  |   at it.unimi.dsi.fastutil.objects.ObjectIterators.unwrap(ObjectIterators.java:228)
infinity_1  |   at it.unimi.dsi.fastutil.objects.ObjectIterators.unwrap(ObjectIterators.java:242)
infinity_1  |   at it.unimi.dsi.fastutil.objects.AbstractObjectCollection.toArray(AbstractObjectCollection.java:83)
infinity_1  |   at org.bukkit.craftbukkit.v1_11_R1.CraftWorld.getLoadedChunks(CraftWorld.java:143)
infinity_1  |   at de.sldk.mc.MetricsController.handle(MetricsController.java:32)
infinity_1  |   at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
infinity_1  |   at org.eclipse.jetty.server.Server.handle(Server.java:564)
infinity_1  |   at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:318)
infinity_1  |   at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
infinity_1  |   at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
infinity_1  |   at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:112)
infinity_1  |   at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
infinity_1  |   at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
infinity_1  |   at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
infinity_1  |   at java.lang.Thread.run(Thread.java:745)

```

Environment:
Using https://hub.docker.com/r/itzg/minecraft-server/  

compose file
```
$ cat ~/config/servy/minecraft/docker-compose.yml
version: '3.0'

services:
  infinity:
    image: itzg/minecraft-server
    restart: always
    volumes:
      - /minecraft_infinity:/data
      - /storage/minecraft_infinity/plugins/dynmap:/data/plugins/dynmap
    ports:
      - 25565:25565
      - 80:8123
      - 9225:9225
    environment:
      - TYPE=SPIGOT
      - CONSOLE=false
      - EULA=TRUE
      - LEVEL_TYPE=LARGEBIOMES
      - PVP=TRUE
      - VIEW_DISTANCE=16
      - MAX_BUILD_HEIGHT=512
      - MAX_PLAYERS=100
      - OPS=zaquestion
      - DIFFICULTY=hard
      - JVM_OPTS=-Xmx8G -Xms1G
```

Things of note:
Server Type Spigot

**Comment by zaquestion** [comment]
Saw more of these today FWIW

**Comment by sladkoff** [comment]
Thanks, I'll look into it!

**Comment by sladkoff** [comment]
I read up on the Bukkit API and I believe the errors occur when accessing the API from a thread other than the main thread (not thread safe).

The embedded Jetty server for serving the metrics endpoint works asynchronously so that's where concurrency errors may happen.

I see two ways to fix this currently:

* Catch and ignore the exceptions (quickest solution, overall the metrics should not be influenced too much if this only happens once in a while...)
* Collect server stats in a scheduled task in a [BukkitRunnable](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/scheduler/BukkitRunnable.html). The metrics endpoint would then expose the latest stats that were available to the scheduled task. That would limit the statistics accuracy to a fixed delay in the exporter and the values parsed by Prometheus would lag behind by a certain amount of time (= scheduler delay).

@zaquestion Can you tell how often those exceptions occur in your case? What's the `scrape_interval` in your Prometheus config? Is the loss of the data points caused by the errors negligible?
I haven't witnessed those on my server yet and also could not reproduce them locally with your docker-compose config. Might have to do with the amount of load your server is getting (I've only been testing with a couple of players + few plugins etc)

**Comment by sladkoff** [comment]
@zaquestion Can you try out [v1.0.1](https://github.com/sladkoff/minecraft-prometheus-exporter/releases/download/v1.0.1/minecraft-prometheus-exporter-1.0.1.jar) and report if you still experience any issues of this kind?
Thank you!

**Comment by sladkoff** [comment]
Closing as this should be fixed in v1.0.1.
Feel free to re-open if problems persist.

**Comment by zaquestion** [comment]
For the record, I did stop seeing those Exceptions.
